### PR TITLE
Test automation harness for testing deduplication rules

### PR DESCRIPTION
**Pull request by nsamadavid**
The tests in the first test suite are written as per requirements for [birth deduplication](https://opencrvs.notion.site/Deduplication-algo-ea397113454b4d35bb77237255231b77). The test suite boots up an elasticsearch container, which is then populated with a sample record. After this we use details of another record to match against the database using our deduplication logic.

### CODE REVIEW

File: packages/search/src/elasticsearch/dbhelper.test.ts [hunk 1]
```diff
@@ -9,30 +9,53 @@
...
   searchForDuplicates,
   searchByCompositionId
 } from '@search/elasticsearch/dbhelper'
 import { mockCompositionBody } from '@search/test/utils'
-import { client } from '@search/elasticsearch/client'
 import { logger } from '@search/logger'
 import { IBirthCompositionBody } from '@search/elasticsearch/utils'
+import * as elasticsearch from '@elastic/elasticsearch'
+import { StartedElasticsearchContainer } from 'testcontainers'
+// @ts-ignore
```
Comment (rikukissa) [inline packages/search/src/elasticsearch/dbhelper.test.ts:24]: Can we somehow get rid of these ts-ignores? TS compiler doesn't seem to complain about these lines without the comments. You can use `yarn test:compilation` to check this.

File: packages/search/src/features/registration/deduplicate/elasticSearchTestContainer.test.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
+/*
```
Comment (rikukissa) [inline packages/search/src/features/registration/deduplicate/elasticSearchTestContainer.test.ts:1]: Can we rename this file to be `deduplication.test.ts`?

File: packages/search/src/features/registration/deduplicate/elasticSearchTestContainer.test.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  container = await startContainer()
+  const host = container?.getHost() ?? '0.0.0.0'
+  const port = container?.getMappedPort(9200) ?? 9200
+
+  client = new elasticsearch.Client({
+    node: `http://${host}:${port}`
+  })
+})
+
+describe('Elastic Search Test Container Automation', () => {
```
Comment (rikukissa) [inline packages/search/src/features/registration/deduplicate/elasticSearchTestContainer.test.ts:37]: Let's change this description to be something more aligned with the fact that we are testing deduplication

**Review by rikukissa — COMMENTED** [review]
Good stuff! A few minor comments. It also seems the CI tests fail to some compilation errors

https://github.com/opencrvs/opencrvs-core/actions/runs/4044754236/jobs/6955311920#step:7:664

File: packages/search/src/elasticsearch/dbhelper.test.ts [hunk 4]
```diff
@@ -9,30 +9,53 @@
...
+    })
+
+    it('should check elasticsearch is up', async () => {
+      await client.ping()
     })
 
     it('should index a composition with proper configuration', async () => {
       indexSpy = jest.spyOn(client, 'index')
-      indexComposition('testId', mockCompositionBody)
+      await indexComposition('testId', mockCompositionBody, client)
```
Comment (rikukissa) [inline packages/search/src/elasticsearch/dbhelper.test.ts:58]: These awaits shouldn't be needed here as the test only checks if the spied function `index` is called but the test doesn't expect any result

File: packages/search/src/features/registration/deduplicate/elasticSearchTestContainer.test.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+jest.setTimeout(10 * 60 * 1000)
+
+let container: StartedElasticsearchContainer
+let client: elasticsearch.Client
+
+beforeAll(async () => {
+  container = await startContainer()
+  const host = container?.getHost() ?? '0.0.0.0'
+  const port = container?.getMappedPort(9200) ?? 9200
```
Comment (rikukissa) [inline packages/search/src/features/registration/deduplicate/elasticSearchTestContainer.test.ts:30]: Do these actually need defaults? I'm thinking if those commands fail, then we probably don't want this client to connect to any other elasticsearch either, right?

File: packages/search/src/elasticsearch/dbhelper.ts [hunk 6]
```diff
@@ -57,7 +63,10 @@ export const updateComposition = async (id: string, body: ICompositionBody) => {
   return response
 }
 
-export const searchForDuplicates = async (body: IBirthCompositionBody) => {
+export const searchForDuplicates = async (
```
Comment (rikukissa) [inline packages/search/src/elasticsearch/dbhelper.ts:66]: Should we remove this function altogether now as it's in the `deduplicate/service.ts`?

File: packages/search/src/features/registration/deduplicate/deduplicate.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  'elasticsearch:7.17.7'
+)
+
+/**
+ * @description Function that starts ElasticSearchTestContainer
+ * @returns {Promise<void>} Promise<void>
+ * **/
+export const startContainer =
+  async (): Promise<StartedElasticsearchContainer> => {
+    return await container
```
Comment (rikukissa) [inline packages/search/src/features/registration/deduplicate/deduplicate.ts:29]: In this case you can only do `return container` as you are returning a promise

File: packages/search/src/elasticsearch/dbhelper.test.ts [hunk 8]
```diff
@@ -9,30 +9,52 @@
  * Copyright (C) The OpenCRVS Authors. OpenCRVS and the OpenCRVS
  * graphic logo are (registered/a) trademark(s) of Plan International.
  */
+// @ts-ignore
```
Comment (rikukissa) [inline packages/search/src/elasticsearch/dbhelper.test.ts:12]: Notice one ts-ignore was left here

File: packages/search/src/features/registration/deduplicate/deduplicate.test.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+describe('Deduplication tests', () => {
+  afterAll(async () => {
+    try {
+      await client.close()
+    } catch (error) {
+    } finally {
+      await stopContainer(container)
+    }
+  })
```
Comment (rikukissa) [inline packages/search/src/features/registration/deduplicate/deduplicate.test.ts:42]: Would it make sense to move this block to the root level of the file, next to beforeAll?

Comment (nsamadavid) [reply packages/search/src/features/registration/deduplicate/deduplicate.test.ts:42]: `beforeAll(async () => {
  container = await startContainer()
  const host = container.getHost()
  const port = container.getMappedPort(9200)

  client = new elasticsearch.Client({
    node: `http://${host}:${port}`
  })
})

afterAll(async () => {
  try {
    await client.close()
  } catch (error) {
  } finally {
    await stopContainer(container)
  }
})`

in this mannere?

Comment (rikukissa) [reply packages/search/src/features/registration/deduplicate/deduplicate.test.ts]: Precisely 👍

**Review by Zangetsu101 — COMMENTED** [review]
This is looking really great!

File: packages/search/src/features/registration/deduplicate/service.ts [hunk 10]
```diff
@@ -37,3 +49,180 @@ const extractRelatesToIDs = (bundle: fhir.Bundle) => {
...
+    bool: {
+      must: [
+        body.childDoB && {
+          range: {
+            childDoB: {
+              gte: subYears(new Date(body.childDoB), 1).toISOString(),
+              lte: addYears(new Date(body.childDoB), 1).toISOString()
+            }
+          }
+        },
```
[the remark this answers is not in the document]
Comment (rikukissa) [reply packages/search/src/features/registration/deduplicate/service.ts]: Changed to 9 months

File: packages/search/src/features/registration/deduplicate/service.ts [hunk 11]
```diff
@@ -37,3 +49,180 @@ const extractRelatesToIDs = (bundle: fhir.Bundle) => {
...
+
+/**
+ * @author
+ * @kind module
+ * @description search for duplicate birth declarations
+ * @param { IBirthCompositionBody } body -  Params
+ * @param client
+ * @returns { Promise<ApiResponse<ISearchResponse<IBirthCompositionBody>, Context>> } -  Promise
+ * **/
+export const searchForDuplicates = async (
```
Comment (naftis) [inline packages/search/src/features/registration/deduplicate/service.ts:61]: great stuff remembering documentation! could we remove the self-documenting parts though; I think TypeScript handles `@param` and `@returns`. Also `@author` and `@kind` are redundant

File: packages/search/src/features/registration/deduplicate/test-util.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+export const ELASTIC_SEARCH_HTTP_PORT = 9200
+
+const container: ElasticsearchContainer = new ElasticsearchContainer(
+  'elasticsearch:7.17.7'
+)
+
+/**
+ * @description Function that starts ElasticSearchTestContainer
+ * @returns {Promise<void>} Promise<void>
+ * **/
```
Comment (naftis) [inline packages/search/src/features/registration/deduplicate/test-util.ts:30]: think `@returns` is redundant here. Let's leave that documentation to TypeScript

**Review by naftis — APPROVED** [review]

**Pull request merged**